// actustime/lib/process/disk-stats.js
'use strict';


var os = require('os');
var childProcess = require('child_process');

/*
 * Sending process related data as metrics.
 */

function DiskStats(agent) {
  this.agent = agent;

  this.headerRegexLinux = undefined;
  this.mountRegexLinux = undefined;

  this.headerRegexOSX = undefined;
  this.mountRegexOSX = undefined;
}
exports.DiskStats = DiskStats;


DiskStats.prototype.init = function() {
  var self = this;

  self.headerRegex = /^(Filesystem)\s+(\w+-blocks)\s+(Used)\s+(Available)\s/;
  self.mountRegex = /^\/dev\/([^\s]+)\s+(\d+)\s+(\d+)\s+(\d+)\s+(\S+)\s+(\S+)/;

  self.agent.timers.setInterval(function() {
    self.collectMetrics();
    self.collectCpuMetrics();
  }, self.agent.interval);

  self.collectMetrics();
};


DiskStats.prototype.collectMetrics = function() {
  var self = this;

  var hostname = os.hostname();

  childProcess.exec('df -Pk', function(err, stdout, stderr) {
    try {
      if(err) return self.agent.logger.error(err);

      var lines = stdout.split('\n');
      if(lines.length == 0) return;

      if(!self.headerRegex.exec(lines[0])) return;

      for(var i = 1; i < lines.length; i++) {
        var mount = {};

        var mountMatch = self.mountRegex.exec(lines[i]);
        if(mountMatch && mountMatch.length >= 5) {
          mount.device = mountMatch[1];
          mount.used = parseInt(mountMatch[3]) / 1024;
          mount.available = parseInt(mountMatch[4]) / 1024;
          mount.percent = mountMatch[5];
          mount.mp = mountMatch[6];
        }

        if(mount.mp && mount.used !== undefined && mount.available !== undefined) {
          var total = mount.used + mount.available;

          self.agent.metric('Disks/' + hostname + '/' + mount.device, 'Total space', total, 'MB', 'set');
          self.agent.metric('Disks/' + hostname + '/' + mount.device, 'Used space', mount.used, 'MB', 'set');
        }
      }
    }
    catch(err) {
      self.agent.logger.error(err);
    }
  });
};

DiskStats.prototype.collectCpuMetrics = function() {
  var self = this;

  var currCpus = os.cpus();

  var currTime = {user:0, nice:0, sys:0, idle:0, irq:0};

  for (var i=0,len=currCpus.length;i<len;i++) {
    var cpu = currCpus[i], total = 0;

    for(var type in cpu.times)
      total += cpu.times[type];

    for(var type in cpu.times) {
      var val = Math.round(100 * cpu.times[type] / total);
      currTime[type] += val;
    }
  }

  for(var type in currTime) {
    currTime[type] = currTime[type]/currCpus.length;
  }

  self.agent.metric('OS', 'CPU user', currTime.user, 'percent', 'avg');
  self.agent.metric('OS', 'CPU nice', currTime.nice, 'percent', 'avg');
  self.agent.metric('OS', 'CPU sys', currTime.sys, 'percent', 'avg');
  self.agent.metric('OS', 'CPU idle', currTime.idle, 'percent', 'avg');
  self.agent.metric('OS', 'CPU irq', currTime.irq, 'percent', 'avg');
};
